cloudinit/ssh_util.py

```diff
@@ -683,3 +701,116 @@ def get_opensshd_upstream_version():
         return upstream_version
     except (ValueError, TypeError):
         LOG.warning("Could not parse sshd version: %s", upstream_version)
+
+
+def _get_early_key_fifo_path(rundir: str) -> pathlib.Path:
+    return pathlib.Path(rundir, "ssh-keygen-finished")
+
+
+def _write_and_close(path: pathlib.Path, data: bytes) -> None:
+    path.write_bytes(data)
+    path.unlink()
+
+
+def _early_generate_host_keys_body(
+    rundir: str, early_key_fifo_path: pathlib.Path
+) -> None:
+    key_dir = get_early_host_key_dir(rundir)
+    key_dir.mkdir(mode=0o600, exist_ok=False)
+
+    for key_type in GENERATE_KEY_NAMES:
+        path = key_dir / (KEY_NAME_TPL % key_type)
+        stdout_path = path.with_suffix(".stdout")
+        stderr_path = path.with_suffix(".stderr")
+        processes = []
+        with open(stdout_path, "w") as stdout, open(
+            stderr_path, "w"
+        ) as stderr:
+            try:
+                # Using subprocess.Popen instead of subp.subp to run
+                # multiple ssh-keygen commands in parallel.
+                p = subprocess.Popen(
```

Comment:
Why generate these in parallel? RSA keygen is time-consuming, but generating the other keys is not, so I'm not convinced that the extra complexity outweighs the benefit.

Reply:
This doesn't really seem any more complex to me, but serial is fine too.